File: Apps/ContactLogoAndroid/app/src/main/java/com/contactlogo/engine/UndoLog.kt
package com.contactlogo.engine

import java.io.File
import java.util.Locale
import java.util.UUID

/**
 * Port of kit `UndoLog` (WRITE POLICY §7 / VISION.md principle 2).
 *
 * Every apply batch first persists the contacts' prior PHOTO bytes (or a
 * "had none" marker) so the batch is one-tap restorable after a relaunch.
 * Bytes live in app-private storage, not Drive.  `allowBackup` is already
 * false; callers still prefer `noBackupFilesDir` so a later backup-flag flip
 * cannot upload prior contact photos.
 *
 * Restore never deletes.  Callers must not retire a batch whose restore has
 * not succeeded — a discarded backup is the one failure that cannot be
 * walked back.
 */
class UndoLog(val directory: File) {

    data class BatchSummary(
        val id: String,
        val createdAt: Double,
        val contactCount: Int
    )

    data class Record(
        val contactId: String,
        val previousImageData: ByteArray?
    )

    data class RestoreOp(
        val contactId: String,
        val previousBytes: ByteArray?
    )

    class UndoException(message: String) : Exception(message)

    internal data class MetaEntry(val contactID: String, val previousImageFile: String?)
    internal data class BatchMeta(val createdAt: Double, val entries: List<MetaEntry>)

    /** Call BEFORE applying.  Returns the batch directory. */
    fun recordBatch(
        entries: List<Record>,
        createdAt: Double = System.currentTimeMillis() / 1000.0
    ): File {
        directory.mkdirs()
        val batchID = UUID.randomUUID().toString()
        val dir = File(directory, batchID)
        if (!dir.mkdirs() && !dir.isDirectory) {
            throw UndoException("could not create undo batch directory")
        }
        val metaEntries = ArrayList<MetaEntry>(entries.size)
        entries.forEachIndexed { index, entry ->
            var file: String? = null
            val previous = entry.previousImageData
            if (previous != null) {
                val name = "previous-$index.img"
                File(dir, name).writeBytes(previous)
                file = name
            }
            metaEntries.add(MetaEntry(entry.contactId, file))
        }
        File(dir, "meta.json").writeText(encodeMeta(BatchMeta(createdAt, metaEntries)))
        return dir
    }

    /**
     * Newest first.  A batch whose meta.json is missing or corrupt is skipped,
     * not thrown, so one bad folder cannot hide the rest of history.
     */
    fun listBatchSummaries(): List<BatchSummary> {
        if (!directory.isDirectory) return emptyList()
        val children = directory.listFiles() ?: return emptyList()
        val out = mutableListOf<BatchSummary>()
        for (child in children) {
            if (!child.isDirectory) continue
            val id = safeComponent(child.name) ?: continue
            val text = try {
                File(child, "meta.json").readText()
            } catch (_: Exception) {
                continue
            }
            val meta = try {
                decodeMeta(text)
            } catch (_: Exception) {
                continue
            }
            out.add(BatchSummary(id, meta.createdAt, meta.entries.size))
        }
        return out.sortedWith(
            compareByDescending<BatchSummary> { it.createdAt }.thenBy { it.id }
        )
    }

    fun prune(keeping: Int = 20, olderThanEpochSeconds: Double? = null) {
        val summaries = listBatchSummaries()
        summaries.forEachIndexed { index, summary ->
            val tooMany = index >= maxOf(0, keeping)
            val cutoff = olderThanEpochSeconds
            val tooOld = cutoff != null && summary.createdAt < cutoff
            if (tooMany || tooOld) {
                try {
                    deleteBatch(summary.id)
                } catch (_: Exception) {
                    // Opportunistic; a stuck folder is retried on the next apply.
                }
            }
        }
    }

    fun deleteBatch(batchID: String) {
        val id = safeComponent(batchID) ?: return
        val dir = File(directory, id)
        if (!dir.exists()) return
        if (!dir.deleteRecursively()) {
            throw UndoException("could not remove undo batch $id")
        }
    }

    /**
     * Reads prior PHOTO bytes for a batch.  Does not mutate the log — a
     * restore that cannot finish must leave this folder on disk.
     */
    fun loadRestoreOps(batchID: String): List<RestoreOp> {
        val id = safeComponent(batchID) ?: throw UndoException("invalid batch id")
        val dir = File(directory, id)
        val metaFile = File(dir, "meta.json")
        if (!metaFile.isFile) throw UndoException("missing meta.json")
        val meta = decodeMeta(metaFile.readText())
        return meta.entries.map { entry ->
            val file = entry.previousImageFile
            if (file != null) {
                // meta.json is on disk, so its file names are untrusted input.
                val name = safeComponent(file)
                    ?: throw UndoException("unsafe previous image name")
                val img = File(dir, name)
                if (!img.isFile) throw UndoException("missing previous image $name")
                RestoreOp(entry.contactID, img.readBytes())
            } else {
                RestoreOp(entry.contactID, null)
            }
        }
    }

    companion object {
        fun safeComponent(raw: String): String? {
            if (raw.isEmpty() || raw.length > 255 || raw == "." || raw == "..") return null
            if (raw.startsWith(".") ||
                raw.contains('/') ||
                raw.contains('\\') ||
                raw.contains('\u0000') ||
                raw.contains("..")
            ) return null
            return raw
        }

        internal fun encodeMeta(meta: BatchMeta): String {
            val entries = meta.entries.joinToString(",") { e ->
                val file = e.previousImageFile?.let { "\"${escapeJson(it)}\"" } ?: "null"
                """{"contactID":"${escapeJson(e.contactID)}","previousImageFile":$file}"""
            }
            return """{"createdAt":${formatEpoch(meta.createdAt)},"entries":[$entries]}"""
        }

        internal fun decodeMeta(text: String): BatchMeta {
            val created = CREATED_RE.find(text)?.groupValues?.get(1)?.toDoubleOrNull()
                ?: throw UndoException("unreadable batch timestamp")
            val entriesBlock = ENTRIES_RE.find(text)?.groupValues?.get(1).orEmpty()
            val entries = ENTRY_RE.findAll(entriesBlock).map { m ->
                val contactID = unescapeJson(m.groupValues[1])
                val file = if (m.groupValues[2] == "null") null else unescapeJson(m.groupValues[3])
                MetaEntry(contactID, file)
            }.toList()
            return BatchMeta(created, entries)
        }

        private val CREATED_RE = Regex("\"createdAt\"\\s*:\\s*([0-9]+(?:\\.[0-9]+)?)")
        private val ENTRIES_RE = Regex("\"entries\"\\s*:\\s*\\[(.*?)]", setOf(RegexOption.DOT_MATCHES_ALL))
        private val ENTRY_RE = Regex(
            "\\{\\s*\"contactID\"\\s*:\\s*\"((?:\\\\.|[^\"\\\\])*)\"\\s*,\\s*\"previousImageFile\"\\s*:\\s*(null|\"((?:\\\\.|[^\"\\\\])*)\")\\s*}"
        )

        private fun formatEpoch(value: Double): String =
            String.format(Locale.US, "%.6f", value)

        private fun escapeJson(s: String): String =
            s.replace("\\", "\\\\").replace("\"", "\\\"")

        private fun unescapeJson(s: String): String =
            s.replace("\\\"", "\"").replace("\\\\", "\\")
    }
}

File: Sources/ContactLogoKit/Store/UndoLog.swift
import Foundation

/// Undo log (WRITE POLICY §7): every apply batch first persists the contacts'
/// prior images (or "had none" markers), so any batch is one-tap restorable —
/// including after a relaunch, which is what `listBatchSummaries()` is for.
public struct UndoLog: Sendable {
    public let directory: URL

    public init(directory: URL? = nil) {
        if let directory {
            self.directory = directory
        } else {
            let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
            self.directory = base.appendingPathComponent("ContactLogo/Undo", isDirectory: true)
        }
    }

    /// One batch as the review UI sees it: newest first, ordered by the
    /// recorded timestamp rather than by the batch's random UUID.
    public struct BatchSummary: Sendable, Identifiable, Equatable {
        public let id: String
        public let createdAt: Date
        public let contactCount: Int

        public init(id: String, createdAt: Date, contactCount: Int) {
            self.id = id
            self.createdAt = createdAt
            self.contactCount = contactCount
        }
    }

    /// meta.json is written and read by us alone. Dates carry sub-second
    /// precision so batches recorded in the same second still sort.
    static func makeEncoder() -> JSONEncoder {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .secondsSince1970
        return encoder
    }

    static func makeDecoder() -> JSONDecoder {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .custom { source in
            let container = try source.singleValueContainer()
            if let seconds = try? container.decode(Double.self) {
                return Date(timeIntervalSince1970: seconds)
            }
            // Batches written by earlier builds used an ISO-8601 string.
            let text = try container.decode(String.self)
            if let date = ISO8601DateFormatter().date(from: text) { return date }
            throw DecodingError.dataCorruptedError(in: container,
                                                   debugDescription: "unreadable batch timestamp")
        }
        return decoder
    }

    struct BatchMeta: Codable {
        struct Entry: Codable {
            let contactID: String
            let previousImageFile: String? // nil → previously no image
        }
        let createdAt: Date
        let entries: [Entry]
    }

    /// A file name is written by us and read back from disk, so both ends are
    /// validated: one path component, no separators, no traversal, no dot
    /// files.  Contact identifiers are never used as file names.
    static func safeComponent(_ raw: String) -> String? {
        guard !raw.isEmpty, raw.count <= 255, raw != ".", raw != ".." else { return nil }
        guard !raw.hasPrefix("."), !raw.contains("/"), !raw.contains("\\"), !raw.contains("\0"),
              !raw.contains("..") else { return nil }
        return raw
    }

    /// Call BEFORE applying. Returns the batch directory.
    @discardableResult
    public func recordBatch(_ entries: [ChangeSet.Entry]) throws -> URL {
        let batchID = UUID().uuidString
        let dir = directory.appendingPathComponent(batchID, isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        var meta: [BatchMeta.Entry] = []
        for (index, entry) in entries.enumerated() {
            var file: String? = nil
            if let previous = entry.previousImageData {
                let name = "previous-\(index).img"
                try previous.write(to: dir.appendingPathComponent(name))
                file = name
            }
            meta.append(.init(contactID: entry.contactID, previousImageFile: file))
        }
        try Self.makeEncoder().encode(BatchMeta(createdAt: Date(), entries: meta))
            .write(to: dir.appendingPathComponent("meta.json"))
        return dir
    }

    /// Chronologically ordered (newest first). A batch whose meta.json is
    /// missing or corrupt is skipped, not thrown.
    public func listBatchSummaries() throws -> [BatchSummary] {
        guard FileManager.default.fileExists(atPath: directory.path) else { return [] }
        let decoder = Self.makeDecoder()
        var out: [BatchSummary] = []
        for url in try FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil)
        where url.hasDirectoryPath {
            guard let id = Self.safeComponent(url.lastPathComponent),
                  let data = try? Data(contentsOf: url.appendingPathComponent("meta.json")),
                  let meta = try? decoder.decode(BatchMeta.self, from: data) else { continue }
            out.append(BatchSummary(id: id, createdAt: meta.createdAt, contactCount: meta.entries.count))
        }
        // Stable: newest first, id as the tie-break so the order is total.
        return out.sorted { a, b in
            a.createdAt == b.createdAt ? a.id < b.id : a.createdAt > b.createdAt
        }
    }

    /// Batch directories, newest first.
    public func listBatches() throws -> [URL] {
        try listBatchSummaries().map { directory.appendingPathComponent($0.id, isDirectory: true) }
    }

    /// Deletes batches beyond the `keeping` most recent, and any batch older
    /// than `olderThan`. Called opportunistically after a successful apply so
    /// prior contact images do not accumulate in Application Support forever.
    public func prune(keeping: Int = 20, olderThan: Date? = nil) throws {
        let summaries = try listBatchSummaries()
        for (index, summary) in summaries.enumerated() {
            let tooMany = index >= max(0, keeping)
            let tooOld = olderThan.map { summary.createdAt < $0 } ?? false
            guard tooMany || tooOld else { continue }
            try? deleteBatch(summary.id)
        }
    }

    /// Permanently removes one batch directory.
    public func deleteBatch(_ batchID: String) throws {
        guard let id = Self.safeComponent(batchID) else { return }
        let dir = directory.appendingPathComponent(id, isDirectory: true)
        guard FileManager.default.fileExists(atPath: dir.path) else { return }
        try FileManager.default.removeItem(at: dir)
    }

    /// Restore a batch: puts previous images back (or removes applied ones).
    public func restore(batchID: String, using provider: any ContactsProvider) async throws {
        guard let id = Self.safeComponent(batchID) else {
            throw CocoaError(.fileNoSuchFile)
        }
        let dir = directory.appendingPathComponent(id, isDirectory: true)
        let metaData = try Data(contentsOf: dir.appendingPathComponent("meta.json"))
        let meta = try Self.makeDecoder().decode(BatchMeta.self, from: metaData)
        for entry in meta.entries {
            if let file = entry.previousImageFile {
                // meta.json is on disk, so its file names are untrusted input.
                guard let name = Self.safeComponent(file) else { continue }
                let data = try Data(contentsOf: dir.appendingPathComponent(name))
                try await provider.setImage(data, forContactID: entry.contactID)
            } else {
                try await provider.removeImage(forContactID: entry.contactID)
            }
        }
    }
}
